transfer-syntax-registry/src/adapters/jpegls.rs

```diff
@@ -93,3 +102,154 @@ impl PixelDataReader for JpegLSAdapter {
+
+        ensure_whatever!(
+            bits_allocated == 8 || bits_allocated == 16,
+            "BitsAllocated other than 8 or 16 is not supported"
+        );
+
+        ensure_whatever!(
+            bits_stored != 1,
+            "BitsStored of 1 is not supported"
+        );
+
+        let bytes_per_sample = (bits_allocated / 8) as usize;
+        let frame_size =
+            cols as usize * rows as usize * samples_per_pixel as usize * bytes_per_sample;
+
+        // identify frame data using the frame index
+        let pixeldata_uncompressed = &src
+            .raw_pixel_data()
+            .context(encode_error::MissingAttributeSnafu { name: "Pixel Data" })?
+            .fragments[0];
+
+        let frame_data = pixeldata_uncompressed
+            .get(frame_size * frame as usize..frame_size * (frame as usize + 1))
+            .whatever_context("Frame index out of bounds")?;
+
+        // Encode the data
+        let mut encoder = CharLS::default();
+
+        let frame_info = FrameInfo {
+            width: cols as u32,
+            height: rows as u32,
+            bits_per_sample: bits_stored as i32,
+            component_count: samples_per_pixel as i32,
+        };
+
+        // prefer lossless encoding by default
+        let quality = options.quality.map(|q| q.clamp(0, 100)).unwrap_or(100);
+
+        // calculate the maximum acceptable error range
+        // based on the requested quality and bit depth
+        let near = ((1 << (bits_stored - 4)) * (100 - quality as i32) / 100).min(4096);
+
+        let compressed_data = encoder
+            .encode(frame_info, near, frame_data)
+            .whatever_context("JPEG-LS encoding failed")?;
+
+        dst.extend_from_slice(&compressed_data);
+
+        let mut changes = if near > 0 {
+            let compressed_frame_size = compressed_data.len();
+
+            let compression_ratio = frame_size as f64 / compressed_frame_size as f64;
+            let compression_ratio = format!("{:.6}", compression_ratio);
+
+            // provide attribute changes
+            vec![
+                // lossy image compression
+                AttributeOp::new(Tag(0x0028, 0x2110), AttributeAction::SetStr("01".into())),
+                // lossy image compression ratio
+                AttributeOp::new(
+                    Tag(0x0028, 0x2112),
+                    AttributeAction::PushStr(compression_ratio.into()),
+                ),
+            ]
+        } else {
+            vec![
+                // lossless image compression
+                AttributeOp::new(Tag(0x0028, 0x2110), AttributeAction::SetIfMissing("00".into())),
+            ]
+        };
+
+        let pmi = src.photometric_interpretation();
+
+        if samples_per_pixel == 1 {
+            // set Photometric Interpretation to Monochrome2
+            // if it was neither of the expected monochromes
+            if pmi != Some("MONOCHROME1") && pmi != Some("MONOCHROME2") {
```

Comment:
Thank you for the review. That is a good catch, though right now I believe no other pixel data writer legitimately has PALETTE COLOR input in mind at the moment. In most cases it would have to be transformed to RGB beforehand. Maybe a utility function somewhere is in order.

At the very least I can stop the encoder from proceeding if the input is palette color. Decoders should prefer relying on whatever is in the encapsulated pixel data bitstream anyway.

Reply:
On GDCM integration there is code to handle this. And change it to RGB would increase the size of the file, the logic to do so should be added to this adapter (as the adapter is changing the PI).

I think it is best approach would to leave the PI as PALETTE_COLOR, and continue if the codec is in Lossless mode. If not, panic with a message indicating that this combination (PALETTE_COLOR + Lossy) should not be used.

Reply:
I am now starting to wonder how decoding should operate here. It's true that if we encode pixel data in Palette Color as is, then decoding it will turn it back to pixel data back in palette color, which may be manageable. But the pixel data reader API as currently written anticipates some changes to PI, so that the pixel data becomes something more usable ([`dicom_encoding::adapters`](https://github.com/Enet4/dicom-rs/blob/master/encoding/src/adapters.rs#L236-L245)).

- for images with 3 channels, it shall decode to RGB
- for monochrome images, it shall either keep MONOCHROME1 or MONOCHROME2

These conversions are important in cases where the PI as declared in the encapsulated pixel data object would either be incompatible with native pixel data or too inconvenient to work with (i.e. `YBR_FULL_422` is a valid PI for native pixel data, but there is little interest in decoding JPEG to something other than RGB).

These PI conversion rules also need to be unambiguous because the output of the pixel data reader is the native pixel data samples (pushed into a `&mut Vec<u8>`) and nothing else.

Thinking out loud, maybe the best way to settle this is adjust one of the rules so that it reads like this:

> - for pixel data with a single sample per pixel, the output shall retain the photometric interpretation declared in the original object if it is one of `MONOCHROME1`, `MONOCHROME2`, or `PALETTE COLOR`. For any other photometric interpretation, the output shall be assumed to be in `MONOCHROME2`.

Reply:
I think in the  PALETTE_COLOR for it to apply, it need to be a lossless encoding, otherwise the image data will not match the color table and the image will not be as expected because the lossy encoding, will change the pixel values.

As for when there are 3 channels, this is where things become interesting, if I am not mistaken, charls works on RGB, so setting the PI as RGB should work fine (I believe this is the case with openjpeg). Now where I have seen issues, is in jpeg as it can be encoded as YBR or RGB. The safest way to do it, I think would be to detect if the stream is YBR or RGB, then convert it to RGB and then set the PI to RGB. Otherwise we would be changing the PI without ensuring that the stream is actually encoded in RGB, and this is when green and pink images start to appear.

Reply:
If we are talking about the encoding process, the approach so far has been to provide the encoded PI on a best-effort basis. A divergence from the real PI is not critical here because decoders are expected to comply with what's in the encapsulated pixel data bitstream. The decoders in the JPEG* family, for instance, already return the pixels of color images in RGB, regardless of the pixel color sampling scheme defined by the bitstream. This means that consumers working with color images can rely on the decoder's output to be in RGB.

For the scenario where `PALETTE COLOR` is the PI of the given raw pixel data, the encoder must either enforce lossless encoding, or bail with an error. In this case we can override the near factor to zero, as either transfer syntax supports lossless encoding ([PS3.5 A.4.3](https://dicom.nema.org/medical/dicom/current/output/chtml/part05/sect_A.4.3.html)).

Reply:
This approach to work with PALETTE_COLOR should work great. It should work fine, maybe we can just document this behavior.